pkg/datasource/opensergo/opensergo_sentinel_assembler.go

```diff
@@ -0,0 +1,252 @@
+	return flowRules
+}
+
+func (assembler OpensergoSentinelRuleAssembler) assembleFlowRulesFromConcurrencyLimitStrategies(pbRules []*faulttolerancePb.FaultToleranceRule, pbStrategyMap map[string]*faulttolerancePb.ConcurrencyLimitStrategy) []flow.Rule {
+	if len(pbRules) == 0 {
+		return []flow.Rule{}
+	}
+	var flowRules []flow.Rule
+	for _, rule := range pbRules {
+		var strategies []*faulttolerancePb.ConcurrencyLimitStrategy
+		for _, strategy := range rule.GetStrategies() {
+			concurrencyLimitStrategy := configkind.ConfigKindRefConcurrencyLimitStrategy{}.GetSimpleName()
+			if strategy.Kind == concurrencyLimitStrategy {
+				strategies = append(strategies, pbStrategyMap[strategy.Name])
+			}
+		}
+		if len(strategies) == 0 {
+			continue
+		}
+
+		for _, targetRef := range rule.GetTargets() {
+			resourceName := targetRef.TargetResourceName
+
+			for _, strategy := range strategies {
+				flowRule := new(flow.Rule)
+				flowRule.Resource = resourceName
+				flowRule = assembler.fillFlowRuleWithConcurrencyLimitStrategy(flowRule, strategy)
+				if flowRule != nil {
+					flowRules = append(flowRules, *flowRule)
+				}
+			}
+		}
+	}
+	return flowRules
+}
+
+func (assembler OpensergoSentinelRuleAssembler) fillFlowRuleWithRateLimitStrategy(flowRule *flow.Rule, pbStrategy *faulttolerancePb.RateLimitStrategy) *flow.Rule {
+	if flowRule == nil || pbStrategy == nil {
+		return flowRule
+	}
+
+	defer func() *flow.Rule {
+		if r := recover(); r != nil {
+			errRecover := errors.Errorf("%+v", r)
+			logging.Error(errRecover, "[OpenSergoDatasource] Ignoring OpenSergo RateLimitStrategy due to covert failure.", "resourceName", flowRule.Resource, "pbStrategy", pbStrategy)
+		}
+		return nil
+	}()
+
+	// TODO fill field-mapping between sentinel-rule and pb-message
+	flowRule.Threshold = float64(pbStrategy.Threshold)
+	flowRule.TokenCalculateStrategy = flow.Direct
+	flowRule.ControlBehavior = flow.Reject
+
+	return flowRule
+}
+
+func (assembler OpensergoSentinelRuleAssembler) fillFlowRuleWithThrottlingStrategy(flowRule *flow.Rule, pbStrategy *faulttolerancePb.ThrottlingStrategy) *flow.Rule {
+	if flowRule == nil || pbStrategy == nil {
+		return flowRule
+	}
+
+	defer func() *flow.Rule {
+		if r := recover(); r != nil {
+			errRecover := errors.Errorf("%+v", r)
+			logging.Error(errRecover, "[OpenSergoDatasource] Ignoring OpenSergo ThrottlingStrategy due to covert failure.", "resourceName", flowRule.Resource, "pbStrategy", pbStrategy)
+		}
+		return nil
+	}()
+
+	// TODO fill field-mapping between sentinel-rule and pb-message
+	flowRule.Threshold = float64(1000 / pbStrategy.MinIntervalMillisOfRequests)
+	flowRule.TokenCalculateStrategy = flow.Direct
+	flowRule.ControlBehavior = flow.Throttling
+	flowRule.MaxQueueingTimeMs = uint32(pbStrategy.QueueTimeoutMillis)
+
+	return flowRule
+}
+
+func (assembler OpensergoSentinelRuleAssembler) fillFlowRuleWithConcurrencyLimitStrategy(flowRule *flow.Rule, pbStrategy *faulttolerancePb.ConcurrencyLimitStrategy) *flow.Rule {
```

Comment:
In Sentinel Go, [the isolation rule](https://pkg.go.dev/github.com/alibaba/sentinel-golang@v1.0.4/core/isolation) is the counterpart of the ConcurrencyLimitStrategy.

Reply:
fixed.